modules/core/src/main/java/org/apache/ignite/internal/util/BasicRateLimiter.java

```diff
@@ -0,0 +1,154 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.util;
+
+import org.apache.ignite.internal.IgniteInterruptedCheckedException;
+import org.apache.ignite.internal.util.typedef.internal.A;
+
+import static java.lang.Math.max;
+import static java.util.concurrent.TimeUnit.MICROSECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * The simplified version of Google Guava smooth rate limiter.<br><br>
+ *
+ * The primary feature of a rate limiter is its "stable rate", the maximum rate that is should
+ * allow at normal conditions. This is enforced by "throttling" incoming requests as needed, i.e.
+ * compute, for an incoming request, the appropriate throttle time, and make the calling thread
+ * wait as much.<br><br>
+ *
+ * The simplest way to maintain a rate of QPS is to keep the timestamp of the last granted
+ * request, and ensure that (1/QPS) seconds have elapsed since then. For example, for a rate of
+ * QPS=5 (5 tokens per second), if we ensure that a request isn't granted earlier than 200ms after
+ * the last one, then we achieve the intended rate. If a request comes and the last request was
+ * granted only 100ms ago, then we wait for another 100ms. At this rate, serving 15 fresh permits
+ * (i.e. for an acquire(15) request) naturally takes 3 seconds.<br><br>
+ *
+ * It is important to realize that such a limiter has a very superficial memory of the past:
+ * it only remembers the last request. if the limiter was unused for a long period of
+ * time, then a request arrived and was immediately granted? This limiter would immediately
+ * forget about that past underutilization.
+ */
+public class BasicRateLimiter {
+    /** Start timestamp. */
+    private final long startTime = System.nanoTime();
+
+    /** Synchronization mutex. */
+    private final Object mux = new Object();
+
+    /**
+     * The interval between two unit requests, at our stable rate. E.g., a stable rate of 5 permits
+     * per second has a stable interval of 200ms.
+     */
+    private double stableIntervalMicros;
+
+    /**
+     * The time when the next request (no matter its size) will be granted. After granting a request,
+     * this is pushed further in the future. Large requests push this further than small requests.
+     */
+    private long nextFreeTicketMicros;
+
+    /**
+     * @param permitsPerSecond Estimated number of permits per second.
+     */
+    public BasicRateLimiter(double permitsPerSecond) {
+        setRate(permitsPerSecond);
+    }
+
+    /**
+     * Updates the stable rate.
+     *
+     * @param permitsPerSecond the new stable rate of this {@code RateLimiter}
```

Comment:
Please fix all comments in this class according to Ignite codestyle (the point at the end, sentence start from the upper case)